File: src/domain/entities/project.ts
import { IssueId } from '../value-objects/issue-id.js';
import { ProjectId } from '../value-objects/project-id.js';
import { ProjectStatus } from '../value-objects/project-status.js';

import type { TimeProvider } from '../interfaces/time-provider.js';

export interface ProjectSnapshot {
  id: string;
  name: string;
  description: string;
  status: string;
  issueIds: string[];
  createdAt: Date;
  updatedAt: Date;
}

export class Project {
  private _id: ProjectId;
  private _name: string;
  private _description: string;
  private _status: string;
  private _issues: IssueId[];
  private _createdAt: Date;
  private _updatedAt: Date;

  private static readonly MAX_ISSUES = 1000;
  private static readonly MAX_NAME_LENGTH = 255;

  constructor(
    id: ProjectId,
    name: string,
    description: string,
    status: string,
    issues: IssueId[],
    createdAt: Date,
    updatedAt: Date,
    private readonly timeProvider?: TimeProvider
  ) {
    this._id = id;
    this._name = this.validateName(name);
    this._description = description;
    this._status = status;
    this._issues = [...issues];
    this._createdAt = new Date(createdAt);
    this._updatedAt = new Date(updatedAt);
  }

  get id(): ProjectId {
    return this._id;
  }

  get name(): string {
    return this._name;
  }

  get description(): string {
    return this._description;
  }

  get status(): string {
    return this._status;
  }

  get issues(): IssueId[] {
    return [...this._issues];
  }

  get createdAt(): Date {
    return new Date(this._createdAt);
  }

  get updatedAt(): Date {
    return new Date(this._updatedAt);
  }

  static create(
    name: string,
    description: string,
    timeProvider?: TimeProvider
  ): Project {
    const id = ProjectId.generate();
    const now = timeProvider?.now() ?? new Date();
    
    return new Project(
      id,
      name,
      description,
      ProjectStatus.Planning,
      [],
      now,
      now,
      timeProvider
    );
  }

  static fromSnapshot(
    snapshot: ProjectSnapshot,
    timeProvider?: TimeProvider
  ): Project {
    const id = ProjectId.from(snapshot.id);
    const issues = snapshot.issueIds.map(issueId => IssueId.from(issueId));
    
    return new Project(
      id,
      snapshot.name,
      snapshot.description,
      snapshot.status,
      issues,
      snapshot.createdAt,
      snapshot.updatedAt,
      timeProvider
    );
  }

  toSnapshot(): ProjectSnapshot {
    return {
      id: this._id.toString(),
      name: this._name,
      description: this._description,
      status: this._status,
      issueIds: this._issues.map(issue => issue.toString()),
      createdAt: this._createdAt,
      updatedAt: this._updatedAt
    };
  }

  addIssue(issueId: IssueId): void {
    if (!this._issues.some(id => id.equals(issueId))) {
      this._issues.push(issueId);
      this.touch();
    }
  }

  removeIssue(issueId: IssueId): void {
    const index = this._issues.findIndex(id => id.equals(issueId));

    if (index !== -1) {
      this._issues.splice(index, 1);
      this.touch();
    }
  }

  hasIssue(issueId: IssueId): boolean {
    return this._issues.some(id => id.equals(issueId));
  }

  issueCount(): number {
    return this._issues.length;
  }

  updateStatus(newStatus: string): void {
    if (this._status === newStatus) {
      return;
    }

    if (!ProjectStatus.canTransition(this._status, newStatus)) {
      throw new Error(`Invalid status transition from ${this._status} to ${newStatus}`);
    }

    this._status = newStatus;
    this.touch();
  }

  canAddIssue(): boolean {
    return this._issues.length < Project.MAX_ISSUES;
  }

  getUnblockedTasks(): IssueId[] {
    // Placeholder implementation - would normally check dependencies
    return [...this._issues];
  }

  isActive(): boolean {
    return ProjectStatus.isActive(this._status);
  }

  isCompleted(): boolean {
    return this._status === ProjectStatus.Completed;
  }

  updateName(name: string): void {
    this._name = this.validateName(name);
    this.touch();
  }

  updateDescription(description: string): void {
    this._description = description;
    this.touch();
  }

  private validateName(name: string): string {
    const trimmedName = name.trim();
    
    if (!trimmedName) {
      throw new Error('Project name cannot be empty');
    }

    if (trimmedName.length >= Project.MAX_NAME_LENGTH) {
      throw new Error('Project name must be less than 255 characters');
    }

    return trimmedName;
  }

  private touch(): void {
    this._updatedAt = this.timeProvider?.now() ?? new Date();
  }
}

File: src/main/kotlin/io/spiralhouse/cycletime/domain/entities/Project.kt
package io.spiralhouse.cycletime.domain.entities

import io.spiralhouse.cycletime.domain.exceptions.DomainException
import io.spiralhouse.cycletime.domain.services.TimeProvider
import io.spiralhouse.cycletime.domain.valueobjects.IssueId
import io.spiralhouse.cycletime.domain.valueobjects.ProjectId
import io.spiralhouse.cycletime.domain.valueobjects.ProjectStatus
import kotlinx.datetime.Instant
import kotlinx.serialization.Serializable

/**
 * Immutable snapshot of a Project entity's state.
 * Used for persistence and serialization without exposing internal mutability.
 */
@Serializable
data class ProjectSnapshot(
    val id: ProjectId,
    val name: String,
    val description: String?,
    val status: ProjectStatus,
    val issues: List<IssueId>,
    val createdAt: Instant,
    val updatedAt: Instant
)

/**
 * Domain entity representing a Project with rich business logic.
 *
 * A Project manages a collection of issues and enforces business rules
 * around status transitions, naming constraints, and issue management.
 *
 * This entity follows Domain-Driven Design principles with:
 * - Factory methods for controlled instantiation
 * - Encapsulated state with business rule enforcement
 * - Snapshot pattern for persistence
 */
class Project private constructor(
    val id: ProjectId,
    private var _name: String,
    private var _description: String?,
    private var _status: ProjectStatus,
    private val _issues: MutableSet<IssueId>,
    val createdAt: Instant,
    private var _updatedAt: Instant,
    private val timeProvider: TimeProvider
) {
    val name: String get() = _name
    val description: String? get() = _description
    val status: ProjectStatus get() = _status
    val issues: List<IssueId> get() = _issues.toList()
    val issueCount: Int get() = _issues.size
    val updatedAt: Instant get() = _updatedAt

    /**
     * Adds an issue to this project.
     *
     * @param issueId The ID of the issue to add
     * @throws DomainException if the project is archived/completed or if the issue already exists
     */
    fun addIssue(issueId: IssueId) {
        if (_status == ProjectStatus.ARCHIVED) {
            throw DomainException("Cannot add issues to archived project")
        }
        if (_status == ProjectStatus.COMPLETED) {
            throw DomainException("Cannot add issues to completed project")
        }
        if (_issues.contains(issueId)) {
            throw DomainException("Issue already exists in project")
        }
        _issues.add(issueId)
        updateTimestamp()
    }

    /**
     * Removes an issue from this project.
     *
     * @param issueId The ID of the issue to remove
     */
    fun removeIssue(issueId: IssueId) {
        _issues.remove(issueId)
        updateTimestamp()
    }

    /**
     * Updates the project name.
     *
     * @param newName The new name for the project
     * @throws DomainException if the name is invalid
     */
    fun updateName(newName: String) {
        validateName(newName)
        _name = newName
        updateTimestamp()
    }

    /**
     * Updates the project description.
     *
     * @param newDescription The new description (can be null)
     */
    fun updateDescription(newDescription: String?) {
        _description = newDescription
        updateTimestamp()
    }

    /**
     * Archives the project, making it read-only.
     */
    fun archive() {
        _status = ProjectStatus.ARCHIVED
        updateTimestamp()
    }

    /**
     * Marks the project as completed.
     */
    fun complete() {
        _status = ProjectStatus.COMPLETED
        updateTimestamp()
    }

    /**
     * Reactivates the project, allowing modifications again.
     */
    fun reactivate() {
        _status = ProjectStatus.ACTIVE
        updateTimestamp()
    }

    /**
     * Creates an immutable snapshot of the current project state.
     *
     * @return A ProjectSnapshot containing the current state
     */
    fun toSnapshot(): ProjectSnapshot {
        return ProjectSnapshot(
            id = id,
            name = _name,
            description = _description,
            status = _status,
            issues = _issues.toList(),
            createdAt = createdAt,
            updatedAt = _updatedAt
        )
    }

    private fun updateTimestamp() {
        _updatedAt = timeProvider.now()
    }

    companion object {
        private const val MAX_NAME_LENGTH = 255
        private const val MIN_NAME_LENGTH = 1

        /**
         * Factory method to create a new Project.
         *
         * @param name The project name (must not be blank, max 255 chars)
         * @param description Optional project description
         * @param timeProvider Time provider for timestamps
         * @return A new Project instance
         * @throws IllegalArgumentException if timeProvider is null
         * @throws DomainException if name validation fails
         */
        fun create(
            name: String,
            description: String?,
            timeProvider: TimeProvider?
        ): Project {
            requireNotNull(timeProvider) { "TimeProvider cannot be null" }
            validateName(name)

            val now = timeProvider.now()
            return Project(
                id = ProjectId.generate(),
                _name = name,
                _description = description,
                _status = ProjectStatus.ACTIVE,
                _issues = mutableSetOf(),
                createdAt = now,
                _updatedAt = now,
                timeProvider = timeProvider
            )
        }

        /**
         * Reconstitutes a Project from a snapshot.
         *
         * @param snapshot The snapshot to reconstitute from
         * @param timeProvider Time provider for timestamps
         * @return A Project instance with the snapshot's state
         * @throws IllegalArgumentException if timeProvider is null
         */
        fun fromSnapshot(
            snapshot: ProjectSnapshot,
            timeProvider: TimeProvider?
        ): Project {
            requireNotNull(timeProvider) { "TimeProvider cannot be null" }

            return Project(
                id = snapshot.id,
                _name = snapshot.name,
                _description = snapshot.description,
                _status = snapshot.status,
                _issues = snapshot.issues.toMutableSet(),
                createdAt = snapshot.createdAt,
                _updatedAt = snapshot.updatedAt,
                timeProvider = timeProvider
            )
        }

        private fun validateName(name: String) {
            when {
                name.isBlank() -> {
                    throw DomainException("Project name cannot be empty")
                }
                name.length > MAX_NAME_LENGTH -> {
                    throw DomainException("Project name cannot exceed $MAX_NAME_LENGTH characters")
                }
            }
        }
    }
}
